bfvmm/include/intrinsics/x86/intel/apic/lapic.h

```diff
@@ -80,86 +179,78 @@ namespace ia32_apic_base
         { bfdebug_subnhex(level, name, get(), msg); }
     }
 
+    inline void dump(int level, std::string *msg = nullptr)
+    {
+        bfdebug_nhex(level, name, get(), msg);
+        bsp::dump(level, msg);
+        extd::dump(level, msg);
+        en::dump(level, msg);
+        apic_base::dump(level, msg);
+    }
+}
+}
+
+namespace apic
+{
+    inline auto present() noexcept
+    {
+        return cpuid::feature_information::edx::apic::is_enabled();
+    }
+
+    inline auto x2apic_supported() noexcept
+    {
+        return cpuid::feature_information::ecx::x2apic::is_enabled();
+    }
+
     namespace state
     {
         constexpr const auto mask = 0xC00ULL;
         constexpr const auto from = 10ULL;
         constexpr const auto name = "state";
 
         constexpr const auto disabled = 0x0ULL;
-        constexpr const auto invalid = 0x4ULL;
-        constexpr const auto xapic = 0x8ULL;
-        constexpr const auto x2apic = 0xCULL;
+        constexpr const auto invalid = 0x1ULL;
+        constexpr const auto xapic = 0x2ULL;
+        constexpr const auto x2apic = 0x3ULL;
 
         inline auto get() noexcept
-        { return get_bits(_read_msr(addr), mask) >> from; }
-
-        inline auto get(value_type msr) noexcept
-        { return get_bits(msr, mask) >> from; }
-
-        inline void set(value_type val) noexcept
-        { _write_msr(addr, set_bits(_read_msr(addr), mask, val << from)); }
+        { return (intel_x64::msrs::ia32_apic_base::get() & mask) >> from; }
 
-        inline auto set(value_type msr, value_type val) noexcept
-        { return set_bits(msr, mask, val << from); }
-
-        inline auto enable_x2apic() noexcept
-        { set(x2apic); }
-
-        inline auto enable_x2apic(value_type msr) noexcept
-        { return set_bits(msr, mask, x2apic << from); }
-
-        inline auto enable_xapic() noexcept
-        { set(xapic); }
-
-        inline auto enable_xapic(value_type msr) noexcept
-        { return set_bits(msr, mask, xapic << from); }
+        inline void disable_apic() noexcept
+        {
+            msrs::ia32_apic_base::extd::disable();
```

Comment:
These really need to be one wrmsr..Basically these govern the lapic state transitions, and so these calls need to be "atomic" so that the caller can query the current state with get() above, and then transition to the state indicated by these functions